### PR TITLE
Abstract CitusHandler and decouple it from configuration

### PR DESCRIPTION
**Pull request by CyberDem0n**
the main issue was that the configuration for Citus handler and for DCS existed in two places, while ideally AbstractDCS should not know many details about what kind of MPP is in use.

To solve the problem we first dynamically create an object implementing AbstractMPP interfaces, which is a configuration for DCS. Later this object is used to instantiate the class implementing AbstractMPPHandler interface.

This is just a starting point, which does some heavy lifting. As a next steps all kind of variables named after Citus in files different from patroni/postgres/mpp/citus.py should be renamed.

In other words this commit takes over the most complex part of #2940, which was never implemented.

### CODE REVIEW

**Comment by zhjwpku** [comment]
@CyberDem0n Can you pls just review my PR #2940 ? The new commit is just the same as my implementation.

File: patroni/dcs/__init__.py [hunk 1]
```diff
@@ -1336,15 +1334,15 @@ class AbstractDCS(abc.ABC):
     _SYNC = 'sync'
     _FAILSAFE = 'failsafe'
 
-    def __init__(self, config: Dict[str, Any]) -> None:
+    def __init__(self, config: Dict[str, Any], mpp: 'AbstractMPP') -> None:
```
Comment (barthisrael) [inline patroni/dcs/__init__.py:1337]: Do we always have to pass a `AbstractMPP` instance?

File: patroni/postgresql/mpp/__init__.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+    group_re: Any  # re.Pattern[str]
+
+    def __init__(self, config: Dict[str, Union[str, int]]) -> None:
+        self._config = config
+
+    def is_enabled(self) -> bool:
+        """Check if MPP is enabled for a given MPP.
+
+        .. note::
+            We just check that the `_config` object isn't empty and expect it do be empty only in case of :class:`Null`.
```
Comment (barthisrael) [inline patroni/postgresql/mpp/__init__.py:25]:
```suggestion
            We just check that the :attr:`_config` object isn't empty and expect it to be empty only in case of :class:`Null`.
```

File: patroni/postgresql/mpp/__init__.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+
+    @property
+    @abc.abstractmethod
+    def group(self) -> Any:
+        """The group for a given MPP implementation."""
+
+    @property
+    @abc.abstractmethod
+    def coordinator_group_id(self) -> Any:
+        """The groupid of the coordinator PostgreSQL cluster."""
```
Comment (barthisrael) [inline patroni/postgresql/mpp/__init__.py:47]:
```suggestion
        """The group id of the coordinator PostgreSQL cluster."""
```

File: patroni/postgresql/mpp/__init__.py [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+
+    def is_coordinator(self) -> bool:
+        """Check whether this node is running in the coordinator PostgreSQL cluster.
+
+        :returns: ``True`` if MPP is enabled and the group id of this node
+                  matches with the coordinator_group_id, otherwise ``False``.
```
Comment (barthisrael) [inline patroni/postgresql/mpp/__init__.py:53]:
```suggestion
                  matches with the :attr:`coordinator_group_id`, otherwise ``False``.
```

File: patroni/postgresql/mpp/__init__.py [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+
+def iter_mpp_classes(
+        config: Optional[Union['Config', Dict[str, Any]]] = None
+) -> Iterator[Tuple[str, Type[AbstractMPP]]]:
+    """Attempt to import MPP modules that are present in the given configuration.
+
+    :param config: configuration information with possible MPP names as keys. If given, only attempt to import MPP
+                   modules defined in the configuration. Else, if ``None``, attempt to import any supported MPP module.
+
+    :returns: an iterator of tuples, each containing the module ``name`` and the imported MPP class object.
```
Comment (barthisrael) [inline patroni/postgresql/mpp/__init__.py:176]:
```suggestion
    :yields: tuples, each containing the module ``name`` and the imported MPP class object.
```

File: patroni/postgresql/mpp/__init__.py [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+        :yields: handler classs for this object.
+        """
+        for cls in self.__class__.__subclasses__():
+            if issubclass(cls, AbstractMPPHandler) and cls.__name__.startswith(self.__class__.__name__):
+                yield cls
+
+    def get_handler_impl(self, postgresql: 'Postgresql') -> 'AbstractMPPHandler':
+        """Find and instantiate Handler implementation of this object.
+
+        :param postgresql: a reference to `Postgresql` object.
```
Comment (barthisrael) [inline patroni/postgresql/mpp/__init__.py:77]:
```suggestion
        :param postgresql: a reference to :class:`Postgresql` object.
```

File: patroni/postgresql/mpp/__init__.py [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+
+        :raises `PatroniException`: if the Handler class haven't been found.
```
Comment (barthisrael) [inline patroni/postgresql/mpp/__init__.py:79]:
```suggestion
        :raises:
            :exc:`PatroniException`: if the Handler class haven't been found.
```

File: patroni/postgresql/mpp/__init__.py [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+
+        :returns: an instantiated class that implements Handler for this object.
+        """
+        for cls in self._get_handler_cls():
+            return cls(postgresql, self._config)
+        raise PatroniException(f'Failed to initialize {self.__class__.__name__}Handler object')
+
+
+class AbstractMPPHandler(AbstractMPP):
```
Comment (barthisrael) [inline patroni/postgresql/mpp/__init__.py:88]: I guess we could include docstrings about what the dev is expected to do when subclassing this. The same comment applies to `AbstractMPP` and other abstract classes / methods.

**Comment by barthisrael** [comment]
I just noticed #2962 wants to merge into this branch. Let me take a look there, some comments here may not make sense after merging 😆

File: patroni/dcs/__init__.py (hunk 1 above)
Comment (CyberDem0n) [reply patroni/dcs/__init__.py:1339]: Yes, we have to.
It needs to get an idea whether we work with just Postgres/Patroni cluster or with MPP (Citus, or similar).

Comment (barthisrael) [reply patroni/dcs/__init__.py:1339]: Right, I forgot to remove this comment, sorry. I noticed the `Null` stuff later.
Thanks for clarifying IAC.

File: patroni/postgresql/mpp/__init__.py [hunk 9]
```diff
@@ -0,0 +1,295 @@
...
+        """
+        AbstractMPPHandler.__init__(self, postgresql, config)
+
+    def handle_event(self, cluster: Cluster, event: Dict[str, Any]) -> None:
+        """Handle an event sent from a worker node.
+
+        :param cluster: the currently known cluster state from DCS.
+        :param event: the event to be handled.
+        """
+        pass
```
Comment (zhjwpku) [inline patroni/postgresql/mpp/__init__.py:220]:
```suggestion
```

File: patroni/postgresql/mpp/__init__.py [hunk 10]
```diff
@@ -0,0 +1,295 @@
...
+
+    def sync_meta_data(self, cluster: Cluster) -> None:
+        """Sync meta data on the coordinator.
+
+        :param cluster: the currently known cluster state from DCS.
+        """
+        pass
```
Comment (zhjwpku) [inline patroni/postgresql/mpp/__init__.py:227]:
```suggestion
```

File: patroni/postgresql/mpp/__init__.py [hunk 11]
```diff
@@ -0,0 +1,295 @@
...
+
+    def on_demote(self) -> None:
+        """On demote handler.
+
+        Is called when the primary was demoted.
+        """
+        pass
```
Comment (zhjwpku) [inline patroni/postgresql/mpp/__init__.py:234]:
```suggestion
```

File: patroni/postgresql/mpp/__init__.py [hunk 12]
```diff
@@ -0,0 +1,295 @@
...
+
+    def schedule_cache_rebuild(self) -> None:
+        """Cache rebuild handler.
+
+        Is called to notify handler that it has to refresh its metadata cache from the database.
+        """
+        pass
```
Comment (zhjwpku) [inline patroni/postgresql/mpp/__init__.py:241]:
```suggestion
```

File: patroni/postgresql/mpp/__init__.py [hunk 13]
```diff
@@ -0,0 +1,295 @@
...
+
+    def bootstrap(self) -> None:
+        """Bootstrap handler.
+
+        Is called when the new cluster is initialized (through ``initdb`` or a custom bootstrap method).
+        """
+        pass
```
Comment (zhjwpku) [inline patroni/postgresql/mpp/__init__.py:248]:
```suggestion
```

File: patroni/postgresql/mpp/__init__.py [hunk 14]
```diff
@@ -0,0 +1,295 @@
...
+
+    def adjust_postgres_gucs(self, parameters: Dict[str, Any]) -> None:
+        """Adjust GUCs in the current PostgreSQL configuration.
+
+        :param parameters: dictionary of GUCs, with key as GUC name and corresponding value as current GUC value.
+        """
+        pass
```
Comment (zhjwpku) [inline patroni/postgresql/mpp/__init__.py:255]:
```suggestion
```

**Review by zhjwpku — CHANGES_REQUESTED** [review]
codacy complain about these unnessary pass statement, remove them to please codacy.

File: patroni/postgresql/mpp/__init__.py [hunk 15]
```diff
@@ -0,0 +1,295 @@
...
+    :param config: configuration information with possible MPP names as keys. If given, only attempt to import MPP
+                   modules defined in the configuration. Else, if ``None``, attempt to import any supported MPP module.
+
+    :yields: tuples, each containing the module ``name`` and the imported MPP class object.
+    """
+    yield from iter_classes(__package__, AbstractMPP, config)
+
+
+def get_mpp(config: Union['Config', Dict[str, Any]]) -> AbstractMPP:
+    """Attempt to load a MPP module from known available implementations.
```
Comment (CyberDem0n) [inline patroni/postgresql/mpp/__init__.py:286]:
```suggestion
    """Attempt to load and instantiate a MPP module from known available implementations.
```

File: patroni/dcs/__init__.py [hunk 16]
```diff
@@ -1359,6 +1357,10 @@ def __init__(self, config: Dict[str, Any]) -> None:
         self._last_failsafe: Optional[Dict[str, str]] = {}
         self.event = Event()
 
+    @property
+    def mpp(self) -> 'AbstractMPP':
```
Comment (barthisrael) [inline patroni/dcs/__init__.py:1361]:
```suggestion
    def mpp(self) -> 'AbstractMPP':
        """Get the effective underlying MPP, if any has been configured."""
```

File: patroni/dcs/__init__.py [hunk 17]
```diff
@@ -1524,7 +1526,7 @@ def is_citus_coordinator(self) -> bool:
 
         :returns: ``True`` if the given node is running as Citus Coordinator (``group=0``).
```
Comment (barthisrael) [inline patroni/dcs/__init__.py:1527]:
```suggestion
        :returns: ``True`` if the given node is running as the MPP Coordinator.
```

File: patroni/dcs/__init__.py [hunk 18]
```diff
@@ -1535,7 +1537,7 @@ def get_citus_coordinator(self) -> Optional[Cluster]:
         :returns: Select :class:`Cluster` instance associated with the Citus Coordinator group ID.
```
Comment (barthisrael) [inline patroni/dcs/__init__.py:1537]:
```suggestion
        :returns: Select :class:`Cluster` instance associated with the MPP Coordinator group ID.
```

File: patroni/postgresql/mpp/__init__.py [hunk 19]
```diff
@@ -0,0 +1,289 @@
...
+                  matches with the :attr:`coordinator_group_id`, otherwise ``False``.
+        """
+        return self.is_enabled() and self.group == self.coordinator_group_id
+
+    def is_worker(self) -> bool:
+        """Check whether this node is running in the coordinator PostgreSQL cluster.
+
+        :returns: ``True`` if MPP is enabled this node is known to be not running
+                  in the coordinator PostgreSQL cluster, otherwise ``False``.
+        """
```
Comment (barthisrael) [inline patroni/postgresql/mpp/__init__.py:74]:
```suggestion
        """Check whether this node is running as a MPP worker PostgreSQL cluster.

        :returns: ``True`` if MPP is enabled and this node is known to be not running
                  as the coordinator PostgreSQL cluster, otherwise ``False``.
        """
```

File: patroni/postgresql/mpp/__init__.py [hunk 20]
```diff
@@ -0,0 +1,289 @@
...
+        return bool(self._config)
+
+    @staticmethod
+    @abc.abstractmethod
+    def validate_config(config: Any) -> bool:
+        """Check whether provided config is good for a given MPP.
+
+        :returns: ``True`` is config passes validation, otherwise ``False``.
+        """
```
Comment (barthisrael) [inline patroni/postgresql/mpp/__init__.py:49]:
```suggestion
        """Check whether provided config is good for a given MPP.

        :param config: configuration of MPP section.

        :returns: ``True`` is config passes validation, otherwise ``False``.
        """
```

File: patroni/postgresql/mpp/citus.py [hunk 21]
```diff
@@ -63,13 +66,31 @@ def __repr__(self) -> str:
...
 
-class CitusHandler(Thread):
+class Citus(AbstractMPP):
 
-    def __init__(self, postgresql: 'Postgresql', config: Optional[Dict[str, Union[str, int]]]) -> None:
-        super(CitusHandler, self).__init__()
+    group_re = re.compile('^(0|[1-9][0-9]*)$')
+
+    @staticmethod
+    def validate_config(config: Union[Any, Dict[str, Union[str, int]]]) -> bool:
```
Comment (barthisrael) [inline patroni/postgresql/mpp/citus.py:74]:
```suggestion
    def validate_config(config: Union[Any, Dict[str, Union[str, int]]]) -> bool:
        """Check whether provided config is good for a given MPP.

        :param config: configuration of ``citus`` MPP section.

        :returns: ``True`` is config passes validation, otherwise ``False``.
        """
```

File: patroni/postgresql/mpp/citus.py [hunk 22]
```diff
@@ -81,18 +102,6 @@ def __init__(self, postgresql: 'Postgresql', config: Optional[Dict[str, Union[st
...
-    def group(self) -> Optional[int]:
-        return int(self._config['group']) if isinstance(self._config, dict) else None
-
-    def is_coordinator(self) -> bool:
-        return self.is_enabled() and self.group() == CITUS_COORDINATOR_GROUP_ID
-
-    def is_worker(self) -> bool:
-        return self.is_enabled() and not self.is_coordinator()
-
     def schedule_cache_rebuild(self) -> None:
```
Comment (barthisrael) [inline patroni/postgresql/mpp/citus.py:105]:
```suggestion
    def schedule_cache_rebuild(self) -> None:
        """Cache rebuild handler.

        Is called to notify handler that it has to refresh its metadata cache from the database.
        """
```

File: patroni/postgresql/mpp/citus.py [hunk 23]
```diff
@@ -388,9 +394,6 @@ def bootstrap(self) -> None:
             conn.close()
 
     def adjust_postgres_gucs(self, parameters: Dict[str, Any]) -> None:
```
Comment (barthisrael) [inline patroni/postgresql/mpp/citus.py:396]:
```suggestion
    def adjust_postgres_gucs(self, parameters: Dict[str, Any]) -> None:
        """Adjust GUCs in the current PostgreSQL configuration.
        
        :param parameters: dictionary of GUCs, with key as GUC name and the corresponding value as current GUC value.
        """
```

File: patroni/postgresql/mpp/citus.py [hunk 24]
```diff
@@ -408,8 +411,7 @@ def adjust_postgres_gucs(self, parameters: Dict[str, Any]) -> None:
         parameters['citus.local_hostname'] = self._postgresql.connection_pool.conn_kwargs.get('host', 'localhost')
 
     def ignore_replication_slot(self, slot: Dict[str, str]) -> bool:
```
Comment (barthisrael) [inline patroni/postgresql/mpp/citus.py:413]:
```suggestion
    def ignore_replication_slot(self, slot: Dict[str, str]) -> bool:
        """Check whether provided replication *slot* existing in the database should not be removed.
        
        .. note::
            MPP database may create replication slots for its own use, for example to migrate data between workers
            using logical replication, and we don't want to suddenly drop them.
            
        :param slot: dictionary containing the replication slot settings, like ``name``, ``database``, ``type``, and
                     ``plugin``.
                     
        :returns: ``True`` if the replication slots should not be removed, otherwise ``False``.
        """
```

**Review by barthisrael — CHANGES_REQUESTED** [review]

File: patroni/ctl.py [hunk 25]
```diff
@@ -340,12 +341,13 @@ def get_dcs(scope: str, group: Optional[int]) -> AbstractDCS:
...
     config.update({'scope': scope, 'patronictl': True})
     if group is not None:
-        config['citus'] = {'group': group}
+        config['citus'] = {'group': group, 'database': 'postgres'}
     config.setdefault('name', scope)
     try:
         dcs = _get_dcs(config)
         if is_citus_cluster() and group is None:
             dcs.is_citus_coordinator = lambda: True
+        click.get_current_context().obj['__mpp'] = dcs.mpp
```
[the remark this answers is not in the document]
Comment (CyberDem0n) [reply patroni/ctl.py:350]: Here were are mutating ctx.obj, and therefore can't use `_get_mpp()`.

File: patroni/postgresql/mpp/__init__.py [hunk 26]
```diff
@@ -0,0 +1,291 @@
+import abc
```
Comment (zhjwpku) [inline patroni/postgresql/mpp/__init__.py:1]:
```suggestion
"""Abstract classes for MPP handler."""
import abc
```

Comment (zhjwpku) [reply patroni/postgresql/mpp/__init__.py:1]: @hughcapet @CyberDem0n pls take a look can this resolving the docstring error?

**Comment by barthisrael** [comment]
👍

**Comment by hughcapet** [comment]
👍
